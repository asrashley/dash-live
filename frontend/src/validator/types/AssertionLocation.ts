export interface AssertionLocation {
    filename: string;
    line: number;
}
