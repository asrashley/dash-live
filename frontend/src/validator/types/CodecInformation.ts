export interface CodecDetails {
    label: string;
    error?: string;
    details: string[];
}

export interface CodecInformation {
    codec: string;
    error?: string;
    details: CodecDetails[];
}
