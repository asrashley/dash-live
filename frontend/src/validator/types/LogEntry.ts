export type LogEntry = {
    id: number;
    level: 'error' | 'info' | 'debug';
    text: string;
};
