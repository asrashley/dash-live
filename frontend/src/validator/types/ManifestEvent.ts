
export interface ManifestEvent {
    text: string[];
}