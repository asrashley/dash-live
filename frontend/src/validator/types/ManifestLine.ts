export interface ManifestLine {
    line: number;
    text: string;
    hasError: boolean;
    errors: string[];
}
