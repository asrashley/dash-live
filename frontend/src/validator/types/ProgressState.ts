export interface ProgressState {
    currentValue?: number;
    minValue: number;
    maxValue: number;
    text: string;
    finished?: boolean;
    error?: boolean;
 }
