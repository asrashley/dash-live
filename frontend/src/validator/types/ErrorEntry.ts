import { AssertionLocation } from "./AssertionLocation";

export interface ErrorEntry {
    assertion: AssertionLocation;
    location: [number, number];
    clause?: string;
    msg: string;
}