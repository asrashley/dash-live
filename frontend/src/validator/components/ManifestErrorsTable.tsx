import { useComputed } from "@preact/signals";
import { UseValidatorWebsocketHook } from "../hooks/useValidatorWebsocket";
import { ErrorEntry } from "../types/ErrorEntry";

function ErrorRow({ assertion, location, msg, clause = "" }: ErrorEntry) {
  const { filename, line } = assertion;
  const [start] = location;
  return (
    <tr>
      <td class="lineno">{start}</td>
      <td class="clause">{clause}</td>
      <td class="message">{msg}</td>
      <td class="source-location">
        {filename}:{line}
      </td>
    </tr>
  );
}

interface ManifestErrorsTableProps {
  errors: UseValidatorWebsocketHook["errors"];
}

export function ManifestErrorsTable({ errors }: ManifestErrorsTableProps) {
  const className = useComputed<string>(() => `${ errors.value.length === 0 ? "d-none": "table table-striped manifest-errors"}`);
  return (
    <table className={className}>
      <thead>
        <tr>
          <th class="lineno">MPD line</th>
          <th class="clause">DASH clause</th>
          <th class="message">Error Details</th>
          <th class="source-location">Source code</th>
        </tr>
      </thead>
      <tbody>
        {errors.value.map((err, idx) => (
          <ErrorRow key={idx} {...err} />
        ))}
      </tbody>
    </table>
  );
}

