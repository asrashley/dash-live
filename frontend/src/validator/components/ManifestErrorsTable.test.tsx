import { signal } from "@preact/signals";
import { describe, expect, test } from "vitest";
import { ErrorEntry } from "../types/ErrorEntry";
import { renderWithProviders } from "../../test/renderWithProviders";
import { ManifestErrorsTable } from "./ManifestErrorsTable";

describe("ManifestErrorsTable component", () => {
  const errors = signal<ErrorEntry[]>([]);

  test("matches snapshot", () => {
    errors.value = [
      {
        assertion: {
          filename: "period.py",
          line: 345,
        },
        location: [20, 30],
        clause: "1.2.3",
        msg: "validator assertion message",
      },
      {
        assertion: {
          filename: "representation.py",
          line: 123,
        },
        location: [120, 130],
        clause: "3.4.5",
        msg: "representation has an error",
      },
    ];
    const { asFragment, getByText, getBySelector } = renderWithProviders(
      <ManifestErrorsTable errors={errors} />
    );
    expect(getBySelector("table").classList.contains("d-none")).toEqual(false);
    errors.value.forEach(({ assertion, msg }) => {
      getByText(`${assertion.filename}:${assertion.line}`);
      getByText(msg);
    });
    expect(asFragment()).toMatchSnapshot();
  });

  test("table is hidden when there are no errors", () => {
    errors.value = [];
    const { getBySelector } = renderWithProviders(
      <ManifestErrorsTable errors={errors} />
    );
    expect(getBySelector("table").classList.contains("d-none")).toEqual(true);
  });
});
