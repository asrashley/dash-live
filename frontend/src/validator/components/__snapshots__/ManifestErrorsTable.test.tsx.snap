// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ManifestErrorsTable component > matches snapshot 1`] = `
<DocumentFragment>
  <table
    class="table table-striped manifest-errors"
  >
    <thead>
      <tr>
        <th
          class="lineno"
        >
          MPD line
        </th>
        <th
          class="clause"
        >
          DASH clause
        </th>
        <th
          class="message"
        >
          Error Details
        </th>
        <th
          class="source-location"
        >
          Source code
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="lineno"
        >
          20
        </td>
        <td
          class="clause"
        >
          1.2.3
        </td>
        <td
          class="message"
        >
          validator assertion message
        </td>
        <td
          class="source-location"
        >
          period.py:345
        </td>
      </tr>
      <tr>
        <td
          class="lineno"
        >
          120
        </td>
        <td
          class="clause"
        >
          3.4.5
        </td>
        <td
          class="message"
        >
          representation has an error
        </td>
        <td
          class="source-location"
        >
          representation.py:123
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
